README
Z(1)                             User Commands                            Z(1)



NAME
       z - jump around

SYNOPSIS
       z [-chlrtx] [regex1 regex2 ... regexn]

AVAILABILITY
       bash, zsh

DESCRIPTION
       Tracks your most used directories, based on 'frequency'.


       After  a  short  learning  phase, z will take you to the most 'frequent'
       directory that matches ALL of the regexes given on the command line, in
       order.

       For example, z foo bar would match /foo/bar but not /bar/foo.


OPTIONS
       -c     restrict matches to subdirectories of the current directory

       -e     echo the best match, don't cd

       -h     show a brief help message

       -l     list only

       -r     match by rank only

       -t     match by recent access only

       -x     remove the current directory from the datafile

EXAMPLES
       z foo         cd to most frequent dir matching foo

       z foo bar     cd to most frequent dir matching foo and bar
       
       z -r foo      cd to highest ranked dir matching foo

       z -t foo      cd to most recently accessed dir matching foo

       z -l foo      list all dirs matching foo (by frecency)

NOTES
   Installation:
       Put something like this in your $HOME/.bashrc or $HOME/.zshrc:

              . /path/to/z.sh

       cd around for a while to build up the db.

       PROFIT!!

       Optionally:
              Set $_Z_CMD to change the command name (default z).
              Set $_Z_DATA to change the datafile (default $HOME/.z).
              Set $_Z_NO_RESOLVE_SYMLINKS to prevent symlink resolution.
              Set $_Z_NO_PROMPT_COMMAND to handle PROMPT_COMMAND/precmd  your-
              self.
              Set $_Z_EXCLUDE_DIRS to an array of directory trees to  exclude.
              Set $_Z_OWNER to allow usage when in 'sudo -s' mode.
              (These  settings  should  go  in  .bashrc/.zshrc before the line
              added above.)
              Install the provided man page z.1  somewhere  in  your  MANPATH,
              like /usr/local/man/man1.

   Aging:
       The rank of directories maintained by z undergoes aging based on a sim-
       ple formula. The rank of each entry is incremented  every  time  it  is
       accessed.  When the sum of ranks is over 9000, all ranks are multiplied
       by 0.99. Entries with a rank lower than 1 are forgotten.

   Frecency:
       Frecency is a portmanteau of 'recent' and 'frequency'. It is a weighted
       rank  that depends on how often and how recently something occurred. As
       far as I know, Mozilla came up with the term.

       To z, a directory that has low ranking but has been  accessed  recently
       will  quickly  have  higher rank than a directory accessed frequently a
       long time ago.

       Frecency is determined at runtime.

   Common:
       When multiple directories match all queries, and they all have a common
       prefix, z will cd to the shortest matching directory, without regard to
       priority.  This has been in effect, if  undocumented,  for  quite  some
       time, but should probably be configurable or reconsidered.

   Tab Completion:
       z  supports tab completion. After any number of arguments, press TAB to
       complete on directories that match each argument. Due to limitations of
       the  completion  implementations,  only  the last argument will be com-
       pleted in the shell.

       Internally, z decides you've requested a completion if the  last  argu-
       ment  passed  is  an  absolute  path to an existing directory. This may
       cause unexpected behavior if the last argument to z begins with /.

ENVIRONMENT
       A function _z() is defined.

       The contents of the variable $_Z_CMD is aliased to _z 2>&1. If not set,
       $_Z_CMD defaults to z.

       The  environment  variable $_Z_DATA can be used to control the datafile
       location. If it is not defined, the location defaults to $HOME/.z.

       The environment variable $_Z_NO_RESOLVE_SYMLINKS can be set to  prevent
       resolving  of  symlinks.  If  it  is  not  set,  symbolic links will be
       resolved when added to the datafile.

       In bash, z appends a command to the PROMPT_COMMAND environment variable
       to maintain its database. In zsh, z appends a function _z_precmd to the
       precmd_functions array.

       The environment variable $_Z_NO_PROMPT_COMMAND can be set if  you  want
       to handle PROMPT_COMMAND or precmd yourself.

       The  environment  variable  $_Z_EXCLUDE_DIRS  can be set to an array of
       directory trees to exclude from tracking.  $HOME  is  always  excluded.
       Directories must be full paths without trailing slashes.

       The  environment  variable  $_Z_OWNER  can  be set to your username, to
       allow usage of z when your sudo environment keeps $HOME set.

FILES
       Data is stored in $HOME/.z. This  can  be  overridden  by  setting  the
       $_Z_DATA  environment variable. When initialized, z will raise an error
       if this path is a directory, and not function correctly.

       A man page (z.1) is provided.

SEE ALSO
       regex(7), pushd, popd, autojump, cdargs

       Please file bugs at https://github.com/rupa/z/



z                                January 2013                             Z(1)


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/rupa/z/tree/HEAD)

[Full Changelog](https://github.com/rupa/z/compare/v1.11...HEAD)

**Closed issues:**

- Typo in README.md [\#222](https://github.com/rupa/z/issues/222)
- Feature Request: Go to next best match? [\#208](https://github.com/rupa/z/issues/208)
- Current stable release version? [\#205](https://github.com/rupa/z/issues/205)
- Suggestion: Z is awesome. It would be great to have a similar command for history. [\#188](https://github.com/rupa/z/issues/188)
- Jump too simple [\#187](https://github.com/rupa/z/issues/187)
- Is the name somehow connected with Charlie the unicorn? [\#167](https://github.com/rupa/z/issues/167)
- Support Symlinked Datafile [\#154](https://github.com/rupa/z/issues/154)
- How to install the c version? [\#135](https://github.com/rupa/z/issues/135)
- Access a directory up to the latest matching regex [\#127](https://github.com/rupa/z/issues/127)

**Merged pull requests:**

- Use `index\(s, t\)` instead of regular expression [\#214](https://github.com/rupa/z/pull/214) ([ericbn](https://github.com/ericbn))

## [v1.11](https://github.com/rupa/z/tree/v1.11) (2017-06-18)
[Full Changelog](https://github.com/rupa/z/compare/v1.10...v1.11)

## [v1.10](https://github.com/rupa/z/tree/v1.10) (2017-06-18)
[Full Changelog](https://github.com/rupa/z/compare/v1.9...v1.10)

**Closed issues:**

- z doesn't record cd [\#202](https://github.com/rupa/z/issues/202)
- z hangs shell on macOS Sierra [\#194](https://github.com/rupa/z/issues/194)
- Fails under zsh [\#192](https://github.com/rupa/z/issues/192)
- Update datafile in background for speed [\#191](https://github.com/rupa/z/issues/191)
- man path update [\#186](https://github.com/rupa/z/issues/186)
- options not work due to unrecognized modifier `1' [\#185](https://github.com/rupa/z/issues/185)
- Project active? [\#183](https://github.com/rupa/z/issues/183)
- \[feature request\] echo the current working directory [\#180](https://github.com/rupa/z/issues/180)
- .z stops updating when the last directory in ~/.z does not exist [\#177](https://github.com/rupa/z/issues/177)
- Can't pipe result of `z -t` [\#176](https://github.com/rupa/z/issues/176)
- Nothing being written to ~/.z [\#175](https://github.com/rupa/z/issues/175)
- Make use of pushd [\#162](https://github.com/rupa/z/issues/162)
- "brew install awk" breaks z [\#150](https://github.com/rupa/z/issues/150)
- z with zsh 4.3.9 not working [\#66](https://github.com/rupa/z/issues/66)

**Merged pull requests:**

- Check intended exit code. Fixes \#177. [\#178](https://github.com/rupa/z/pull/178) ([paulp](https://github.com/paulp))

## [v1.9](https://github.com/rupa/z/tree/v1.9) (2015-09-13)
[Full Changelog](https://github.com/rupa/z/compare/v1.8...v1.9)

**Closed issues:**

- License? [\#160](https://github.com/rupa/z/issues/160)
- List entries with tab delimiter [\#156](https://github.com/rupa/z/issues/156)
- not creating .z in archlinux [\#149](https://github.com/rupa/z/issues/149)
- z not creating .z file in zsh [\#147](https://github.com/rupa/z/issues/147)
- alias path='echo -e ${PATH//:/\\n}' breaks z [\#144](https://github.com/rupa/z/issues/144)
- z - [\#142](https://github.com/rupa/z/issues/142)
- error 'sed: 1: extra characters at the end of d command' when using -x option [\#141](https://github.com/rupa/z/issues/141)
- z acts like in zsh when ran from tmux [\#140](https://github.com/rupa/z/issues/140)
- .z isn't created [\#139](https://github.com/rupa/z/issues/139)
- Permission denied [\#134](https://github.com/rupa/z/issues/134)
- z -x sed error on OS X [\#125](https://github.com/rupa/z/issues/125)
- Exclude wildcard [\#124](https://github.com/rupa/z/issues/124)
- array plusequal use breaks old bash [\#122](https://github.com/rupa/z/issues/122)
- Can get it to install [\#120](https://github.com/rupa/z/issues/120)
- typo: $tmpfile should be $tempfile [\#115](https://github.com/rupa/z/issues/115)
- \_Z\_NO\_RESOLVE\_SYMLINKS isn't recognized on OS X 10.8.4 [\#114](https://github.com/rupa/z/issues/114)
- z appears to not recognize a dot as a literal dot [\#113](https://github.com/rupa/z/issues/113)
- terminal hangs when sourcing bashrc [\#112](https://github.com/rupa/z/issues/112)
- How to source z.sh in eshell? [\#108](https://github.com/rupa/z/issues/108)
- I get ERROR: z.sh is meant to be sourced... whenever I run a shell script [\#107](https://github.com/rupa/z/issues/107)
- stray temporary files [\#84](https://github.com/rupa/z/issues/84)
- Can't install z with zsh version 5.0.0 with oh-my-zsh [\#77](https://github.com/rupa/z/issues/77)
- Can't install z on zsh 4.3.12 \(x86\_64-redhat-linux-gnu\) [\#67](https://github.com/rupa/z/issues/67)
- Try to find an approximate match by popping the pattern list [\#40](https://github.com/rupa/z/issues/40)
- Windows support? [\#5](https://github.com/rupa/z/issues/5)

**Merged pull requests:**

- Fix special char escaping so `mawk` likes it [\#152](https://github.com/rupa/z/pull/152) ([somasis](https://github.com/somasis))
- commit c507bdd broke zsh -- this fixes it [\#148](https://github.com/rupa/z/pull/148) ([tomhoover](https://github.com/tomhoover))
- Allow usage of z while in 'sudo -s' mode [\#137](https://github.com/rupa/z/pull/137) ([dan-bolsun](https://github.com/dan-bolsun))
- use command pwd instead of pwd [\#132](https://github.com/rupa/z/pull/132) ([lge88](https://github.com/lge88))
- Prevent extraneous shifting [\#130](https://github.com/rupa/z/pull/130) ([notwa](https://github.com/notwa))
- redirect to /dev/null instead of '-q' option for solaris 10 or earlier [\#109](https://github.com/rupa/z/pull/109) ([uchida](https://github.com/uchida))

## [v1.8](https://github.com/rupa/z/tree/v1.8) (2013-07-07)
[Full Changelog](https://github.com/rupa/z/compare/v1.7...v1.8)

**Closed issues:**

- Errors when sshing to account which is sourcing z.sh [\#106](https://github.com/rupa/z/issues/106)
- PROMPT\_COMMAND error [\#105](https://github.com/rupa/z/issues/105)

zz.sh
#!/bin/bash

# pure shell version of 'new j'
# has frecency, rank, and recent
# has common and case preference
# lists to stderr
# protect multiple PROMPT_COMMAND appends

zz() {
 local datafile=$HOME/.z
 if [ "$1" = "--add" ]; then
  # add
  shift
  # $HOME isn't worth matching
  [ "$*" = "$HOME" ] && return
  awk -v q="$*" -v t="$(date +%s)" -F"|" '
   BEGIN { l[q] = 1; d[q] = t }
   $2 >= 1 {
    if( $1 == q ) {
     l[$1] = $2 + 1
     d[$1] = t
    } else {
     l[$1] = $2
     d[$1] = $3
    }
    count += $2
   }
   END {
    if( count > 1000 ) {
     for( i in l ) print i "|" 0.9*l[i] "|" d[i] # aging
    } else for( i in l ) print i "|" l[i] "|" d[i]
   }
  ' $datafile 2>/dev/null > $datafile.tmp
  mv -f $datafile.tmp $datafile
 elif [ "$1" = "--complete" ]; then
  # tab completion
  awk -v q="$2" -F"|" '
   BEGIN { split(substr(q,3),a," ") }
   {
    if( system("test -d \"" $1 "\"") ) next
    for( i in a ) $1 !~ a[i] && $1 = ""; if( $1 ) print $1
   }
  ' $datafile 2>/dev/null
 else
  # list/go
  while [ "$1" ]; do case $1 in
   -h) echo "zz [-h][-l][-r][-t] args" >/dev/stderr; return;;
   -l) local list=1;;
   -r) local typ="rank";;
   -t) local typ="recent";;
   --) while [ "$1" ]; do shift; local q="$q $1";done;;
    *) local q="$q $1";;
  esac; local n=$1; shift; done
  [ "$q" ] || local list=1
  # if we hit enter on a completion just go there
  [ -d "$n" ] && {
   cd "$n"
   return
  }
  cd="$(awk -v t="$(date +%s)" -v list="$list" -v typ="$typ" -v args="$q" -F"|" '
   function frecent(rank, time) {
    dx = t-time
    if( dx < 3600 ) return rank*4
    if( dx < 86400 ) return rank*2
    if( dx < 604800 ) return rank/2
    return rank/4
   }
   function output(r, s, c) {
    if( list ) {
     if( typ == "recent" ) {
      cmd = "sort -nr >/dev/stderr"
     } else cmd = "sort -n >/dev/stderr"
     for( i in r ) if( r[i] ) printf "%-15s %s\n", r[i], i | cmd
     if( c ) printf "%-15s %s\n", "common:", c > "/dev/stderr"
    } else {
     if( c ) s = c
     print s
    }
   }
   function common(r, a, nc) {
    for( i in r ) {
     if( ! r[i] ) continue
     if( !shortest || length(i) < length(shortest) ) shortest = i
    }
    if( shortest == "/" ) return
    for( i in r ) {
     if( ! r[i] ) continue
     if( i !~ shortest ) x = 1
    }
    if( x ) return
    if( nc ) {
     for( i in a ) if( tolower(shortest) !~ tolower(a[i]) ) x = 1
    } else for( i in a ) if( shortest !~ a[i] ) x = 1
    if( !x ) return shortest
   }
   BEGIN { split(args,a," ") }
   {
    if( system("test -d \"" $1 "\"") ) next
    if( typ == "rank" ) {
     f = $2
    } else if( typ == "recent" ) {
     f = t-$3
    } else f = frecent($2, $3)
    case[$1] = nocase[$1] = f
    for( i in a ) if( $1 !~ a[i] ) delete case[$1]
    for( i in a ) if( tolower($1) !~ tolower(a[i]) ) delete nocase[$1]
    if( case[$1] > oldf ) {
     cx = $1
     oldf = case[$1]
    } else if( nocase[$1] > noldf ) {
     ncx = $1
     noldf = nocase[$1]
    }
    print $0 >> FILENAME ".tmp"
   }
   END {
    if( cx ) {
     output(case, cx, common(case, a, 0))
    } else if( ncx ) {
     output(nocase, ncx, common(nocase, a, 1))
    }
   }
  ' $datafile)"
  mv -f $datafile.tmp $datafile
  [ "$cd" ] && cd "$cd"
 fi
}
# tab completion
complete -C 'zz --complete "$COMP_LINE"' zz
# populate directory list. avoid clobbering other PROMPT_COMMANDs.
echo $PROMPT_COMMAND | grep -q "zz" || {
 PROMPT_COMMAND='zz --add "$(pwd -P)";'"$PROMPT_COMMAND"
}


z_prompt_command.c
/* PROMPT_COMMAND for z.sh */

#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <time.h>
#include <unistd.h>

int main(int argc, char **argv) {

    static const float MULTIPLIER = 0.99;
    static const int RANK_SUM_MAX = 6000;

    char * _Z_DATA;
    char datafile[PATH_MAX];
    char tmpfile[PATH_MAX];
    char agedfile[PATH_MAX];
    FILE *read;
    FILE *write;
    FILE *aged;
    char line[LINE_MAX];
    char pth[PATH_MAX];
    float rank;
    int timestamp;
    struct stat s;
    time_t epoch = time(NULL);
    int found = 0;
    float rank_sum = 0;

    (void)argc;
    if (argv[1] == NULL) return EXIT_FAILURE;

    /* Don't track $HOME */
    if (strcmp(argv[1], getenv("HOME")) == 0) return EXIT_SUCCESS;

    _Z_DATA = getenv("_Z_DATA");
    if( _Z_DATA == NULL ) {
        sprintf(datafile, "%s/.z", getenv("HOME"));
    } else {
        sprintf(datafile, "%s", _Z_DATA);
    }
    read = fopen(datafile, "r");
    if (read == NULL) return EXIT_FAILURE;

    srand(time(NULL));

    sprintf(tmpfile, "%s.%d", datafile, rand());
    write = fopen(tmpfile, "w");
    if (write == NULL) return EXIT_FAILURE;

    sprintf(agedfile, "%s.%d", datafile, rand());
    aged = fopen(agedfile, "w");
    if (aged == NULL) return EXIT_FAILURE;

    while (fgets(line, sizeof line, read) != NULL) {
        sscanf(line, "%[^|]|%f|%d\n", pth, &rank, &timestamp);
        if (stat(pth, &s) == 0 && s.st_mode & S_IFDIR ) {
            rank_sum += rank;
            if (strcmp(pth, argv[1]) == 0) {
                fprintf(write, "%s|%f|%d\n", pth, rank + 1, (int) epoch);
                fprintf(aged, "%s|%f|%d\n", pth, rank * 1, (int) epoch);
                found = 1;
            } else if (rank >= 1) {
                fprintf(write, "%s", line);
                fprintf(aged, "%s|%f|%d\n", pth, rank * MULTIPLIER, timestamp);
            }
        }
    }
    if (!found) {
        fprintf(write, "%s|1|%d\n", argv[1], (int) epoch);
        fprintf(aged, "%s|1|%d\n", argv[1], (int) epoch);
    }

    fclose(read);
    fclose(write);
    fclose(aged);

    if (rank_sum > RANK_SUM_MAX) {
        if (rename(agedfile, datafile) != 0) return EXIT_FAILURE;
        if (unlink(tmpfile) != 0) return EXIT_FAILURE;
    } else {
        if (rename(tmpfile, datafile) != 0) return EXIT_FAILURE;
        if (unlink(agedfile) != 0) return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}


z.sh
# Copyright (c) 2009 rupa deadwyler. Licensed under the WTFPL license, Version 2

# maintains a jump-list of the directories you actually use
#
# INSTALL:
#     * put something like this in your .bashrc/.zshrc:
#         . /path/to/z.sh
#     * cd around for a while to build up the db
#     * PROFIT!!
#     * optionally:
#         set $_Z_CMD in .bashrc/.zshrc to change the command (default z).
#         set $_Z_DATA in .bashrc/.zshrc to change the datafile (default ~/.z).
#         set $_Z_NO_RESOLVE_SYMLINKS to prevent symlink resolution.
#         set $_Z_NO_PROMPT_COMMAND if you're handling PROMPT_COMMAND yourself.
#         set $_Z_EXCLUDE_DIRS to an array of directories to exclude.
#         set $_Z_OWNER to your username if you want use z while sudo with $HOME kept
#
# USE:
#     * z foo     # cd to most frecent dir matching foo
#     * z foo bar # cd to most frecent dir matching foo and bar
#     * z -r foo  # cd to highest ranked dir matching foo
#     * z -t foo  # cd to most recently accessed dir matching foo
#     * z -l foo  # list matches instead of cd
#     * z -c foo  # restrict matches to subdirs of $PWD

[ -d "${_Z_DATA:-$HOME/.z}" ] && {
    echo "ERROR: z.sh's datafile (${_Z_DATA:-$HOME/.z}) is a directory."
}

_z() {

    local datafile="${_Z_DATA:-$HOME/.z}"

    # if symlink, dereference
    [ -h "$datafile" ] && datafile=$(readlink "$datafile")

    # bail if we don't own ~/.z and $_Z_OWNER not set
    [ -z "$_Z_OWNER" -a -f "$datafile" -a ! -O "$datafile" ] && return

    _z_dirs() {
        local line
        while read line; do
            # only count directories
            [ -d "${line%%\|*}" ] && echo "$line"
        done < "$datafile"
        return 0
    }

    # add entries
    if [ "$1" = "--add" ]; then
        shift

        # $HOME isn't worth matching
        [ "$*" = "$HOME" ] && return

        # don't track excluded directory trees
        local exclude
        for exclude in "${_Z_EXCLUDE_DIRS[@]}"; do
            case "$*" in "$exclude*") return;; esac
        done

        # maintain the data file
        local tempfile="$datafile.$RANDOM"
        awk < <(_z_dirs 2>/dev/null) -v path="$*" -v now="$(date +%s)" -F"|" '
            BEGIN {
                rank[path] = 1
                time[path] = now
            }
            $2 >= 1 {
                # drop ranks below 1
                if( $1 == path ) {
                    rank[$1] = $2 + 1
                    time[$1] = now
                } else {
                    rank[$1] = $2
                    time[$1] = $3
                }
                count += $2
            }
            END {
                if( count > 9000 ) {
                    # aging
                    for( x in rank ) print x "|" 0.99*rank[x] "|" time[x]
                } else for( x in rank ) print x "|" rank[x] "|" time[x]
            }
        ' 2>/dev/null >| "$tempfile"
        # do our best to avoid clobbering the datafile in a race condition.
        if [ $? -ne 0 -a -f "$datafile" ]; then
            env rm -f "$tempfile"
        else
            [ "$_Z_OWNER" ] && chown $_Z_OWNER:$(id -ng $_Z_OWNER) "$tempfile"
            env mv -f "$tempfile" "$datafile" || env rm -f "$tempfile"
        fi

    # tab completion
    elif [ "$1" = '--complete' -a -s "$datafile" ]; then
        _z_dirs | awk -v q="$2" -F'|' '
            BEGIN {
                q = substr(q, 3)
                # smart-case
                if( q == tolower(q) ) imatch = 1
                gsub(/ /, ".*", q)
            }
            {
                if( imatch ) {
                    if( tolower($1) ~ q ) print $1
                } else if( $1 ~ q ) print $1
            }
        ' 2>/dev/null

    else
        # list/go
        while [ "$1" ]; do case "$1" in
            --) while [ "$1" ]; do shift; local fnd="$fnd${fnd:+ }$1";done;;
            -*) local opt=${1:1}; while [ "$opt" ]; do case ${opt:0:1} in
                    c) local fnd="^$PWD $fnd";;
                    e) local echo=echo;;
                    h) echo "${_Z_CMD:-z} [-cehlrtx] args" >&2; return;;
                    l) local list=1;;
                    r) local typ="rank";;
                    t) local typ="recent";;
                    x) sed -i -e "\:^${PWD}|.*:d" "$datafile";;
                esac; opt=${opt:1}; done;;
             *) local fnd="$fnd${fnd:+ }$1";;
        esac; local last=$1; [ "$#" -gt 0 ] && shift; done
        [ "$fnd" -a "$fnd" != "^$PWD " ] || local list=1

        # if we hit enter on a completion just go there
        case "$last" in
            # completions will always start with /
            /*) [ -z "$list" -a -d "$last" ] && builtin cd "$last" && return;;
        esac

        # no file yet
        [ -f "$datafile" ] || return

        local cd
        cd="$( < <( _z_dirs ) awk -v t="$(date +%s)" -v list="$list" -v typ="$typ" -v q="$fnd" -F"|" '
            function frecent(rank, time) {
                # relate frequency and time
                dx = t - time
                if( dx < 3600 ) return rank * 4
                if( dx < 86400 ) return rank * 2
                if( dx < 604800 ) return rank / 2
                return rank / 4
            }
            function output(matches, best_match, common) {
                # list or return the desired directory
                if( list ) {
                    cmd = "sort -n >&2"
                    for( x in matches ) {
                        if( matches[x] ) {
                            printf "%-10s %s\n", matches[x], x | cmd
                        }
                    }
                    if( common ) {
                        printf "%-10s %s\n", "common:", common > "/dev/stderr"
                    }
                } else {
                    if( common ) best_match = common
                    print best_match
                }
            }
            function common(matches) {
                # find the common root of a list of matches, if it exists
                for( x in matches ) {
                    if( matches[x] && (!short || length(x) < length(short)) ) {
                        short = x
                    }
                }
                if( short == "/" ) return
                for( x in matches ) if( matches[x] && index(x, short) != 1 ) {
                    return
                }
                return short
            }
            BEGIN {
                gsub(" ", ".*", q)
                hi_rank = ihi_rank = -9999999999
            }
            {
                if( typ == "rank" ) {
                    rank = $2
                } else if( typ == "recent" ) {
                    rank = $3 - t
                } else rank = frecent($2, $3)
                if( $1 ~ q ) {
                    matches[$1] = rank
                } else if( tolower($1) ~ tolower(q) ) imatches[$1] = rank
                if( matches[$1] && matches[$1] > hi_rank ) {
                    best_match = $1
                    hi_rank = matches[$1]
                } else if( imatches[$1] && imatches[$1] > ihi_rank ) {
                    ibest_match = $1
                    ihi_rank = imatches[$1]
                }
            }
            END {
                # prefer case sensitive
                if( best_match ) {
                    output(matches, best_match, common(matches))
                } else if( ibest_match ) {
                    output(imatches, ibest_match, common(imatches))
                }
            }
        ')"

        [ $? -eq 0 ] && [ "$cd" ] && {
          if [ "$echo" ]; then echo "$cd"; else builtin cd "$cd"; fi
        }
    fi
}

alias ${_Z_CMD:-z}='_z 2>&1'

[ "$_Z_NO_RESOLVE_SYMLINKS" ] || _Z_RESOLVE_SYMLINKS="-P"

if type compctl >/dev/null 2>&1; then
    # zsh
    [ "$_Z_NO_PROMPT_COMMAND" ] || {
        # populate directory list, avoid clobbering any other precmds.
        if [ "$_Z_NO_RESOLVE_SYMLINKS" ]; then
            _z_precmd() {
                (_z --add "${PWD:a}" &)
            }
        else
            _z_precmd() {
                (_z --add "${PWD:A}" &)
            }
        fi
        [[ -n "${precmd_functions[(r)_z_precmd]}" ]] || {
            precmd_functions[$(($#precmd_functions+1))]=_z_precmd
        }
    }
    _z_zsh_tab_completion() {
        # tab completion
        local compl
        read -l compl
        reply=(${(f)"$(_z --complete "$compl")"})
    }
    compctl -U -K _z_zsh_tab_completion _z
elif type complete >/dev/null 2>&1; then
    # bash
    # tab completion
    complete -o filenames -C '_z --complete "$COMP_LINE"' ${_Z_CMD:-z}
    [ "$_Z_NO_PROMPT_COMMAND" ] || {
        # populate directory list. avoid clobbering other PROMPT_COMMANDs.
        grep "_z --add" <<< "$PROMPT_COMMAND" >/dev/null || {
            PROMPT_COMMAND="$PROMPT_COMMAND"$'\n''(_z --add "$(command pwd '$_Z_RESOLVE_SYMLINKS' 2>/dev/null)" 2>/dev/null &);'
        }
    }
fi
